src/base/emu-i386/simx86/interp.c

```diff
@@ -622,9 +620,17 @@ static unsigned int _Interp86(unsigned int PC)
  * returns data (could be modified), so compiled code can write it back
  */
 unsigned int Sim_helper(unsigned int mem_ref, unsigned int data, int mode,
-			uint32_t *flags, unsigned int opc, unsigned int arg)
+			uint32_t *flags, unsigned int opc, unsigned int arg,
+			unsigned int P0)
 {
 	unsigned int temp;
+	if (opc != IRET) {
+			// save for emu_pagefault_handler(),
+			// except for IRET where it's not used from
+			// here and TheCPU.eip contains the popped eip.
```

Comment:
Maybe revert the patch that
puts tmp value to eip too early?
I remember it was giving some
unification between PM and RM,
but now this...

Reply:
it's complicated, as CS is also changed.

I think in the end Sim_helper is a lot like the cpatch functions which all get passed eip.
So it may be best, instead of using `__builtin_return_address` to have the asm code pass eip as the 7th parameter,
instead of P0, as that would be similar to what cpatch is doing already, for consistency. That would bring me back to #2609 then.